### PR TITLE
Beryllium lnwallet 20220414

### PR DESCRIPTION
**Pull request by eoliveros**
- wrap the content of the pages in an alert window.
  - lightning getinfo
  - new address
  - transaction history
  - (all) pages
- withdrawal/psbt:
  - the option to send sat or btc
  - showing the balance in both sat and btc
- sign psbt:
  - shows on-chain balance in btc and sat
- broadcast psbt:
  - shows on-chain balance in btc and sat
- create invoice:
  - checks the amount is not "0"
  - ln-channel balance is now showing
- pay invoice:
    - ln-channel balance is now showing
- invoice history:
  - deleted this page from index
- payment history:
  - deleted this page from index
  - removed route
- open channel:
  - shows on-chain balance in sat
- list forwards:
  - deleted this page from index
- lightning transaction history:
  - combined the invoice and payment history in a series rather than table.
  - shows the ln-channel balance

### CODE REVIEW

File: ln.py [hunk 1]
```diff
@@ -116,7 +116,9 @@ def list_invoices(self):
             description = invoice["description"]
             payment_hash = invoice["payment_hash"]
             expires_at = invoice["expires_at"]
-            amount_msat = invoice["amount_msat"]
+            amount_msat = 0
+            if amount_msat:
```
Comment (djpnewton) [inline ln.py:120]: this condition will never be true (```amount_msat = 0```)

File: ln.py [hunk 2]
```diff
@@ -128,10 +130,42 @@ def list_invoices(self):
             if status == 'paid':
                 pay_index = invoice["pay_index"]
                 amount_received_msat = invoice["amount_received_msat"]
+                amount_received_sats = _msat_to_sat(invoice["amount_received_msat"])
```
Comment (djpnewton) [inline ln.py:133]:
```suggestion
                amount_received_sats = _msat_to_sat(amount_received_msat)
```

File: ln.py [hunk 3]
```diff
@@ -128,10 +130,42 @@ def list_invoices(self):
...
+            if bolt11:
+                bolt11 = sendpay["bolt11"]
+            groupid = None
+            if groupid:
+                groupid = sendpay["groupid"]
+            payment_preimage = None
+            fees_sats = None
+            if status == "complete":
+                paid_at = created_at
+                paid_date = datetime.datetime.fromtimestamp(paid_at, pytz.timezone('Pacific/Auckland'))
```
Comment (djpnewton) [inline ln.py:165]: we dont want the timezone hardcoded here, we should use the server timezone

File: ln_wallet_endpoint.py [hunk 4]
```diff
@@ -52,6 +52,7 @@ def list_txs():
     """ Returns template of on-chain txs """
     rpc = LnRpc()
     transactions = rpc.list_txs()
+    logger.info('txs: %s', transactions)
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:55]: remove this?

File: ln_wallet_endpoint.py [hunk 5]
```diff
@@ -121,12 +122,12 @@ def rebalance_channel():
         # I'm assuming there will only be one channel for each node, but I'm
         # using an array in case there's more
         peers[i]["channel_states"] = []
+        can_send_total = 0
```
Comment (djpnewton) [inline ln_wallet_endpoint.py]: seems to be unused?

File: ln_wallet_endpoint.py [hunk 6]
```diff
@@ -121,12 +122,12 @@ def rebalance_channel():
...
         for channel in peers[i]["channels"]:
             if channel["state"] == 'CHANNELD_NORMAL':
                 peers[i]["sats_total"] += int(channel["msatoshi_total"]) / 1000
-                peers[i]["can_send"] += int(channel["msatoshi_to_us"]) / 1000
-                peers[i]["can_receive"] += int(
-                    channel["out_msatoshi_fulfilled"]) / 1000
+                peers[i]["can_send"] += int(channel["spendable_msatoshi"]) / 1000
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:128]: use ```_msat_to_sat``` function

File: ln_wallet_endpoint.py [hunk 7]
```diff
@@ -135,7 +136,7 @@ def rebalance_channel():
         peers[i]["sats_total"] = int(peers[i]["sats_total"])
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:135]: these lines seem redundant

File: ln_wallet_endpoint.py [hunk 8]
```diff
@@ -155,23 +156,27 @@ def pay_invoice():
...
-def list_received_transactions():
+def lightning_transactions():
     """ Returns received transactions """
+    dict_txs = []
+    unsorted_list_dict = []
     rpc = LnRpc()
     received_txs = rpc.list_invoices()
-    return render_template("lightning/received_transactions.html", received_txs=received_txs)
+    send_txs = rpc.list_sendpays()
+    funds_dict=rpc.list_funds()
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:169]: fix spacing around ```=```

File: ln_wallet_endpoint.py [hunk 9]
```diff
@@ -198,14 +203,17 @@ def channel_opener():
...
+    return render_template('lightning/channel_opener.html', funds_dict=funds_dict)
 
 @ln_wallet.route('/create_psbt', methods=['GET', 'POST'])
 @roles_accepted(Role.ROLE_ADMIN)
 def create_psbt():
     """ Returns template for creating a PSBT """
     rpc = LnRpc()
     onchain = int(rpc.list_funds()["sats_onchain"]) / 100000000
+    onchain_sats = int(rpc.list_funds()["sats_onchain"])
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:215]:
```suggestion
    onchain_sats = int(rpc.list_funds()["sats_onchain"])
    onchain = onchain_sats / 100000000
```

File: ln_wallet_endpoint.py [hunk 10]
```diff
@@ -235,11 +245,14 @@ def create_psbt():
...
     return render_template(
-        'lightning/create_psbt.html', onchain=onchain, addrs=addrs, amounts=amounts, mode=mode, psbt=psbt)
+        'lightning/create_psbt.html', onchain=onchain, addrs=addrs, amounts=amounts, mode=mode, psbt=psbt, onchain_sats=onchain_sats)
 
 @ln_wallet.route('/sign_psbt', methods=['GET', 'POST'])
 @roles_accepted(Role.ROLE_ADMIN)
 def sign_psbt():
+    rpc = LnRpc()
+    onchain = int(rpc.list_funds()["sats_onchain"]) / 100000000
+    onchain_sats = int(rpc.list_funds()["sats_onchain"])
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:254]:
```suggestion
    onchain_sats = int(rpc.list_funds()["sats_onchain"])
    onchain = onchain_sats / 100000000
```

File: ln_wallet_endpoint.py [hunk 11]
```diff
@@ -250,11 +263,14 @@ def sign_psbt():
...
             flash(f'Sign failed: {e}', 'danger')
-    return render_template('lightning/sign_psbt.html', signed_psbt=signed_psbt)
+    return render_template('lightning/sign_psbt.html', signed_psbt=signed_psbt, onchain=onchain, onchain_sats=onchain_sats)
 
 @ln_wallet.route('/broadcast_psbt', methods=['GET', 'POST'])
 @roles_accepted(Role.ROLE_ADMIN)
 def broadcast():
+    rpc = LnRpc()
+    onchain = int(rpc.list_funds()["sats_onchain"]) / 100000000
+    onchain_sats = int(rpc.list_funds()["sats_onchain"])
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:272]:
```suggestion
    onchain_sats = int(rpc.list_funds()["sats_onchain"])
    onchain = onchain_sats / 100000000
```

File: templates/lightning/create_psbt.html [hunk 12]
```diff
@@ -35,7 +53,7 @@ <h2>Create PSBT/Withdrawal</h2>
         <div class="input-group">
             <input type="text" id="amount" name="amount" class="payment form-control" placeholder="Amount" required>
             <div class="input-group-append">
-                <button type="button" id="max" class="btn btn-sm">Max</button>
+		<button class="btn btn-outline-secondary" type="button" id="max">Max</button>
```
Comment (djpnewton) [inline templates/lightning/create_psbt.html:56]: fix spacing

File: templates/lightning/create_psbt.html [hunk 13]
```diff
@@ -61,7 +80,7 @@ <h2>Create PSBT/Withdrawal</h2>
         }
         bootbox.confirm({
             title: `Confirm Create ${action}`,
-            message: `Total: ${totalBtc} BTC. Do you want to create the ${action}?`,
+		message: `Total: ${totalBtc}${unit}. Do you want to create the ${action}?`,
```
Comment (djpnewton) [inline templates/lightning/create_psbt.html:83]: fix spacing

File: templates/lightning/create_psbt.html [hunk 14]
```diff
@@ -105,7 +124,13 @@ <h2>Create PSBT/Withdrawal</h2>
         $('#max').click(function() {
             var ele = $('#max').parent().prev();
             console.log(ele);
-            ele.val('{{ onchain }}');
+            const unit = document.querySelector('input[name="unit"]:checked').value;
+            if (unit == 'btc')
+                return ele.val('{{ onchain }}');
+            else if (unit == 'sat')
+		return ele.val('{{ onchain_sats }}');
```
Comment (djpnewton) [inline templates/lightning/create_psbt.html:131]: fix spacing

File: templates/lightning/index.html [hunk 15]
```diff
@@ -1,17 +1,18 @@
...
 {% block content %}
+<a class="btn btn-secondary" href="/"><i class="glyphicon glyphicon-chevron-left"></i> Back</a>&nbsp
 <br/>
 <h2>BALANCES</h2>
 <div class='some-page-wrapper'>
     <div class='row'>
         <div class='dashboard'>
         <p>on-chain balance<p>
-        <p><center>{{ funds_dict.sats_onchain }} sats</center></p>
+	<p><center>{{ funds_dict.sats_onchain }} <strong>SAT</strong></center></p>
```
Comment (djpnewton) [inline templates/lightning/index.html:11]: fix spacing

File: templates/lightning/index.html [hunk 16]
```diff
@@ -1,17 +1,18 @@
...
         </div>
         <div class='dashboard'>
         <p>ln-channel balance<p>
-        <p><center>{{ funds_dict.sats_channels }} sats</center></p>
+	<p><center>{{ funds_dict.sats_channels }} <strong>SAT</strong></center></p>
```
Comment (djpnewton) [inline templates/lightning/index.html:15]: fix spacing

File: templates/lightning/layout.html [hunk 17]
```diff
@@ -3,4 +3,10 @@
 {% block header %}
 <!-- Socket.IO JS -->
 <script src="{{ url_for('static',filename='assets/lightning/socket.io.js') }}"></script>
-{% endblock %}
\ No newline at end of file
+{% endblock %}
+
+{% block content %}
```
Comment (djpnewton) [inline templates/lightning/layout.html:8]: this layout template should not have a content block

File: templates/lightning/lightning_transactions.html [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+<h2>Lightning Transaction History</h2>
+These are the transactions recorded by the wallet.
+<br>
+<p align="right">{{ record_no }} transactions recorded</p>
+    {% for tx in sorted_txs %}
+        <div class="alert alert-dismissible alert-info">
+            <button class={% if tx.amount_sent_sats is defined %}"btn btn-danger disabled"{% elif tx.amount_received_sats is defined %}"btn btn-success disabled"{% endif %} type="button" data-toggle="collapse" data-target="#collapseExample{{loop.index}}" aria-expanded="false" aria-controls="collapseExample">
+                {{tx.paid_date}}
+            </button>
+	    <b>{% if tx.amount_sent_sats is defined %}-{{tx.amount_sent_sats}}{% elif tx.amount_received_sats is defined %}+{{tx.amount_received_sats}}{% endif %} SAT{% if tx.amount_sent_sats is defined%}(fees included){% endif %}</b>
```
Comment (djpnewton) [inline templates/lightning/lightning_transactions.html:19]: fix spacing

File: templates/lightning/lightning_transactions.html [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+                <div class="card card-body">
+                    {% if tx.amount_sent_sats is defined %}
+                        <b>Amount:</b> {{tx.amount_sent_sats}} SAT<br>
+                        <b>Destination:</b> {{tx.destination}}<br>
+                    {% elif tx.amount_received_sats is defined %}
+                        <b>Amount:</b> {{tx.amount_received_sats}} SAT<br>
+                        {% if tx.description is defined %}
+                            <b>Description:</b> {{tx.description}}<br> 
+                        {% endif %}
+			<b>Label:</b> {{tx.label}}<br>
```
Comment (djpnewton) [inline templates/lightning/lightning_transactions.html:30]: fix spacing

File: templates/lightning/lightning_transactions.html [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+                    {% endif %}
+		    <b>Payment hash:</b> {{tx.payment_hash}}<br>
```
Comment (djpnewton) [inline templates/lightning/lightning_transactions.html:32]: fix spacing

File: templates/lightning/layout.html (hunk 17 above)
Comment (djpnewton) [reply templates/lightning/layout.html:8]: you should put a new block in the main layout template called ```navigation-bar``` or something

**Review by djpnewton — CHANGES_REQUESTED** [review]

File: ln.py (hunk 1 above)
Comment (eoliveros) [reply ln.py:120]: reverted back to line 119 when there is no amount_msat = 0 and condition and getting an error:
```
app_1                       |     received_txs = rpc.list_invoices()
app_1                       |   File "/app/ln.py", line 119, in list_invoices
app_1                       |     amount_msat = invoice["amount_msat"]
app_1                       | KeyError: 'amount_msat'
```

will look at it further and see if i can figure this out.

File: ln.py (hunk 2 above)
Comment (eoliveros) [reply ln.py:133]: changed.

File: ln_wallet_endpoint.py (hunk 5 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py]: deleted this in my other commit.

File: ln_wallet_endpoint.py (hunk 4 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py:55]: removed

File: ln.py (hunk 1 above)
Comment (eoliveros) [reply ln.py:120]: since this is transaction related, changed it.

File: ln.py (hunk 3 above)
Comment (eoliveros) [reply ln.py:165]: removing the pytz.timezone would result on everything being defaulted to UTC.
will most likely need to write a function that would make a os.system to query the timezone being used by the server. need to investigate further.

File: ln_wallet_endpoint.py (hunk 6 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py:128]: changed to msat_to_sat

File: ln_wallet_endpoint.py (hunk 7 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py:135]: removed.

File: ln_wallet_endpoint.py (hunk 8 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py:169]: fixed

File: ln_wallet_endpoint.py (hunk 9 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py:215]: changed.

File: ln_wallet_endpoint.py (hunk 10 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py:254]: changed.

File: templates/lightning/create_psbt.html (hunk 13 above)
Comment (eoliveros) [reply templates/lightning/create_psbt.html:83]: fixed

File: templates/lightning/create_psbt.html (hunk 14 above)
Comment (eoliveros) [reply templates/lightning/create_psbt.html:131]: changed the spacing

File: templates/lightning/index.html (hunk 15 above)
Comment (eoliveros) [reply templates/lightning/index.html:11]: fixed.

File: templates/lightning/index.html (hunk 16 above)
Comment (eoliveros) [reply templates/lightning/index.html:15]: fixed

File: templates/lightning/lightning_transactions.html (hunk 18 above)
Comment (eoliveros) [reply templates/lightning/lightning_transactions.html:19]: fixed

File: templates/lightning/lightning_transactions.html (hunk 20 above)
Comment (eoliveros) [reply templates/lightning/lightning_transactions.html:32]: fixed.

File: templates/lightning/lightning_transactions.html (hunk 19 above)
Comment (eoliveros) [reply templates/lightning/lightning_transactions.html:30]: fixed

File: templates/lightning/create_psbt.html (hunk 12 above)
Comment (eoliveros) [reply templates/lightning/create_psbt.html:56]: fixed

File: templates/lightning/layout.html (hunk 17 above)
Comment (eoliveros) [reply templates/lightning/layout.html:8]: this is the result when i add a navbar:
image

its basically creating one new navbar on top of the main navbar. not good.

Comment (eoliveros) [reply templates/lightning/layout.html:8]: ill delete the button from the lightning/layout and put it on individual pages instead

File: ln.py (hunk 3 above)
Comment (djpnewton) [reply ln.py:164]: we can set the timezone in the container if we want to

File: ln.py [hunk 21]
```diff
@@ -126,12 +124,48 @@ def list_invoices(self):
             payment_preimage = None
             status = invoice["status"]
             if status == 'paid':
+                amount_msat = invoice["amount_msat"]
+                amount_sats = _msat_to_sat(amount_msat)
                 pay_index = invoice["pay_index"]
                 amount_received_msat = invoice["amount_received_msat"]
+                amount_received_sats = _msat_to_sat(amount_received_msat)
                 paid_at = invoice["paid_at"]
+                #paid_date = datetime.datetime.fromtimestamp(paid_at)
```
Comment (djpnewton) [inline ln.py:133]: dead code

File: ln.py [hunk 22]
```diff
@@ -126,12 +124,48 @@ def list_invoices(self):
...
                 paid_date = datetime.datetime.fromtimestamp(paid_at, pytz.timezone('Pacific/Auckland'))
```
Comment (djpnewton) [inline ln.py:134]: remove hard coded timezone

File: ln.py [hunk 23]
```diff
@@ -126,12 +124,48 @@ def list_invoices(self):
...
+                bolt11 = sendpay["bolt11"]
+            groupid = None
+            if groupid:
+                groupid = sendpay["groupid"]
+            payment_preimage = None
+            fees_sats = None
+            if status == "complete":
+                paid_at = created_at
+                paid_date = datetime.datetime.fromtimestamp(paid_at, pytz.timezone('Pacific/Auckland'))
+                #paid_date = datetime.datetime.fromtimestamp(paid_at)
```
Comment (djpnewton) [inline ln.py:165]: dead code

File: ln_wallet_endpoint.py [hunk 24]
```diff
@@ -198,14 +197,17 @@ def channel_opener():
             flash(Markup(f'successfully added node id: {node_id[0]} with the amount: {amount}'), 'success')
         except Exception as e: # pylint: disable=broad-except
             flash(Markup(e.args[0]), 'danger')
-    return render_template('lightning/channel_opener.html')
+    rpc = LnRpc()
+    funds_dict=rpc.list_funds()
+    return render_template('lightning/channel_opener.html', funds_dict=funds_dict)
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:202]:
```suggestion
    return render_template('lightning/channel_opener.html', funds_dict=LnRpc().list_funds())
```

File: ln_wallet_endpoint.py [hunk 25]
```diff
@@ -155,23 +149,28 @@ def pay_invoice():
...
+    send_txs = rpc.list_sendpays()
+    funds_dict = rpc.list_funds()
+    dict_txs.append(received_txs)
+    dict_txs.append(send_txs)
+    for unsorted_dict in dict_txs[0]:
+        unsorted_list_dict.append(unsorted_dict)
+    for unsorted_dict in dict_txs[1]:
+        unsorted_list_dict.append(unsorted_dict)
+    sorted_txs = sorted(unsorted_list_dict, key=lambda d: d["paid_at"], reverse=True)
+    #sorted_txs = sorted(unsorted_list_dict, key=lambda d: d["paid_date"], reverse=True)
```
Comment (djpnewton) [inline ln_wallet_endpoint.py:171]: dead code

**Review by djpnewton — CHANGES_REQUESTED** [review]

File: ln.py (hunk 3 above)
Comment (eoliveros) [reply ln.py:164]: trying the following in ln.py:
```
if 'SERVER_TIMEZONE' in os.environ:
    server_timezone = os.environ['SERVER_TIMEZONE']
else:
    server_timezone = "UTC"
```

added a line in .env file for SERVER_TIMEZONE

when checking the list_transactions page, its always coming up as UTC timezone.

ive tried adding app.config in app_core as well. its not working though,

File: ln.py (hunk 21 above)
Comment (eoliveros) [reply ln.py]: removed the dead code

File: ln.py (hunk 22 above)
Comment (eoliveros) [reply ln.py]: should be changeable via env variable.

File: ln.py (hunk 3 above)
Comment (eoliveros) [reply ln.py]: changed so this can be configured via env variable

File: ln.py (hunk 23 above)
Comment (eoliveros) [reply ln.py]: removed the dead code

File: ln_wallet_endpoint.py (hunk 24 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py]: removed line 200 and 201. changed line 202 as well.

File: ln_wallet_endpoint.py (hunk 25 above)
Comment (eoliveros) [reply ln_wallet_endpoint.py]: deleted dead code.

**Comment by djpnewton** [comment]
it will be easier to remove pytz from the code and just use this method (https://confluence.atlassian.com/kb/how-to-set-the-timezone-for-docker-container-976780914.html) that should automatically update all the places we format dates

**Review by eoliveros — COMMENTED** [review]
should be resolved.

**Comment by eoliveros** [comment]
> it will be easier to remove pytz from the code and just use this method (https://confluence.atlassian.com/kb/how-to-set-the-timezone-for-docker-container-976780914.html) that should automatically update all the places we format dates

not sure how to do that.

**Comment by djpnewton** [comment]
> not sure how to do that. The part that sets the date/time is using
> 
> ```
> datetime.datetime.fromtimestamp(timestamp)
> ```
> 
> will default to UTC. thats why using the pytz.timezone is necessary. Unless you are aware of some other way.
> 
> also, theres suddenly a conflict on this push not sure why. This is repo is getting way too big to redo.

https://docs.python.org/3/library/datetime.html#datetime.datetime.fromtimestamp

```If optional argument tz is None or not specified, the timestamp is converted to the platform’s local date and time```

So putting ```TZ=XXX``` in the .env file (see https://confluence.atlassian.com/kb/how-to-set-the-timezone-for-docker-container-976780914.html) should affect every place we do a timestamp conversion using the default tz=None, this is much easier then adding a manual timezone in a hundred different places in the code

**Comment by eoliveros** [comment]
> `If optional argument tz is None or not specified, the timestamp is converted to the platform’s local date and time`

```
- added TZ in .env
- added {TZ} in docker-compose.yml
- removed the pytz module in ln.py'
- changed to timestamp({timestamp})
```

TESTED on my local instance and appears to work so i pushed the changes.

File: templates/lightning/broadcast_psbt.html [hunk 26]
```diff
@@ -1,11 +1,11 @@
 {% extends "lightning/layout.html" %}
 {% block title %}Create Invoice{% endblock %}
+{% block navigation %}
```
Comment (djpnewton) [inline templates/lightning/broadcast_psbt.html:3]: you should not need this

File: docker-compose.yml [hunk 27]
```diff
@@ -45,6 +45,7 @@ services:
       WINDCAVE_MOCK: ${WINDCAVE_MOCK}
       EXCHANGE_ACCOUNT_MOCK: ${EXCHANGE_ACCOUNT_MOCK}
       REGISTRATION_DISABLE: ${REGISTRATION_DISABLE}
+      TZ: ${TZ}
```
[the remark this answers is not in the document]
Comment (eoliveros) [reply docker-compose.yml:48]: Yes, its for setting up the timezone on the server.

**Comment by djpnewton** [comment]
i fixed the merge conflicts

**Pull request merged**